pysteps/verification/salscores.py

```diff
@@ -0,0 +1,345 @@
+# -- coding: utf-8 --
+"""
+pysteps.verification.salscores
+==============================
+
+The Spatial-Amplitude-Location (SAL) score.
+
+.. autosummary::
+    :toctree: ../generated/
+
+    sal
+    sal_structure
+    sal_amplitude
+    sal_location
+"""
+from math import sqrt, hypot
+
+import numpy as np
+from scipy.ndimage.measurements import center_of_mass
+
+from pysteps.exceptions import MissingOptionalDependency
+from pysteps.feature import tstorm as tstorm_detect
+
+try:
+    import pandas as pd
+
+    pandas_imported = True
+except ImportError:
+    pandas_imported = False
+
+try:
+    from skimage.measure import regionprops_table
+
+    skimage_imported = True
+except ImportError:
+    skimage_imported = False
+
+
+def sal(
+    prediction,
+    observation,
+    tstorm_kwargs=None,
+):
+    """
+    Compute the Structure-Amplitude-Location (SAL) spatial verification metric.
+
+    Parameters
+    ----------
+    prediction: array-like
+        Array of shape (m,n) with prediction data.
+    observation: array-like
+        Array of shape (m,n)  with observation data.
+    tstorm_kwargs: dict, optional
+        Optional dictionary containing keyword arguments for the tstorm feature
+        detection algorithm.
+        See the documentation of :py:mod:`pysteps.feature.tstorm`.
+
+    Returns
+    -------
+    sal: tuple of floats
+        A tuple of floats containing the structure, amplitude, location components of SAL.
+
+    References
+    ----------
+    :cite: Wernli, H., Hofmann, C., & Zimmer, M. (2009).
+    :cite: Feldmann, M., Germann, U., Gabella, M., & Berne, A. (2021).
+
+    Notes
+    -----
+    This implementation uses the thunderstorm detection algorithm by Feldmann et al (2021)
+    for the identification of precipitation objects within the considered domain.
+
+    This approach uses multi-threshold algorithm to detect objects, instead of having a
+    single threshold (f).
+
+    See also
+    --------
+    pysteps.feature.tstorm
+    """
+    if np.nanmax(observation >= 0.1) & np.nanmax(
+        prediction >= 0.1
+    ):  # to avoid errors of nan values or very low precipitation
```

Comment:
I'm not convinced by this ... can we do it better? 

I guess the presence of precipitation is problematic for Structure and Location, but not really for Amplitude? Can we make the code robust enough? Or at list make the threshold an optional argument ...

Reply:
I agree it is not the best solution. Maybe If there is no precipitation in observation or forecast, just show a message "No object is detected in observation/ forecast"?

Reply:
> I agree it is not the best solution. Maybe If there is no precipitation in observation or forecast, just show a message "No object is detected in observation/ forecast"?

mmmh I wonder if it should not up to the end user to filter out cases without enough precipitation. We should always return a value when possible, and maybe NaN if no cells are found for those components that depend on object detection (structure and location).

> Also, I think we need to mention that scikit-image package should be 0.18 or higher since the name of some functions are different in older versions.

Yep I realized that: the properties names of `regionprops` changed with v0.19 (see https://github.com/scikit-image/scikit-image/releases). I tried to use the old names (max_intensity, intensity_image), since these are supposed to be still supported in higher versions, but the last test failed and I still need to have a look into it...

Reply:
I agree.

I also tried using the old names but it didn't work. Using higher versions (>=0.18) and the new names solved my problem.

Reply:
Good to know eheh, then I might revert back to the new names and add a check for the sklearn version